libs/constrained-generators/src/Constrained/Base.hs

```diff
@@ -6084,3 +6080,132 @@ nubOrdMemberSpec message xs =
         , "The input is the empty list."
         ]
     )
+
+-- ==============================================================================
+-- Experimental changes to conformsToSpecM
+
+runTermE :: Env -> Term fn a -> Either (NE.NonEmpty String) a
+runTermE env = \case
+  Lit a -> Right a
+  V v -> case lookupEnv env v of
+    Just a -> Right a
+    Nothing -> Left (pure ("Couldn't find " ++ show v ++ " in " ++ show env))
+  App f ts -> do
+    vs <- mapMList (fmap Identity . runTermE env) ts
+    pure $ uncurryList_ runIdentity (sem f) vs
+
+checkPredsE ::
+  FunctionLike fn =>
+  NE.NonEmpty String ->
+  Env ->
+  [Pred fn] ->
+  Maybe (NE.NonEmpty String)
+checkPredsE msgs env ps =
+  -- and <$> mapM (checkPred env) ps
+  case catMaybes (fmap (checkPredE env msgs) ps) of
+    [] -> Nothing
+    (x : xs) -> Just (NE.nub (foldr (<>) x xs))
```

Comment:
What is the reason for putting `x` at the end? Was that intentional?
I'd expected something like this instead:
```suggestion
    xs -> Just (NE.nub (mconcat xs))
```

```haskell
ghci> foldr (<>) "x" ["foo", "bar"]
"foobarx"
ghci> mconcat ["x", "foo", "bar"]
"xfoobar"
```


Same comment applies to all other places where `foldr` is used in the same manner

Reply:
The problem is that NE.NonEmpty, does not have a an empty list like thing.
So one has to pattern match on (x : xs) because if the input is the empty list, there is no way to handle things. So the only thing to do is pattern match against (x : xs)
and then handle things in this strange way.

Reply:
I think you missed what I was asking. I understand the part about the non empty. The question is about concatenation and placing the first element last while doing that

Reply:
For one thing, NonEmpty does not have a Monoid instance.  But the comment about the order is bothersome.

Reply:
I suppose we could use foldl
ghci>  foldl (<>) "x" ["foo", "bar"] 
"xfoobar"